helm-app-operator/pkg/helm/helm.go

```diff
@@ -442,8 +465,9 @@ func (c installer) tillerRendererForCR(r *unstructured.Unstructured) *tiller.Rel
 	return tiller.NewReleaseServer(env, internalClientSet, false)
 }
 
-func releaseName(r *unstructured.Unstructured) string {
-	return fmt.Sprintf("%s-%s", operatorName, r.GetName())
+func getReleaseName(r *unstructured.Unstructured) string {
+	shortUID := strings.Replace(string(r.GetUID()), "-", "", -1)[:8]
```

Comment:
Although perhaps unlikely we could still get collisions with just the last 8 chars of the UUID.
I think some parts of the UUID are usually constant based on the implementation.

We could reduce those odds by base64 encoding the UUID from 32 chars to 22 chars.
E.g https://gist.github.com/hasbro17/201d727ca6402c067dd399a8a059c32c
based on https://medium.com/@ablab/url-safe-uuid-encoding-in-go-4ba3b4ce4206

Or hashing the 32 chars(minus 4 hyphens) down to 8 or 10 chars. That might be a little safer than just lopping off the last 8 chars.

So we can explore the best option but at least add a TODO for now on how we could improve the short UID generation.

Reply:
Looks like Kubernetes uses UUID v1, which uses the a combination of the clock and MAC address of the host:
* https://github.com/kubernetes/apimachinery/blob/master/pkg/util/uuid/uuid.go#L33
* https://github.com/pborman/uuid/blob/master/version1.go#L11-L23

I spun up a deployment with two replicas and compared their UIDs:
```
$ kubectl get pods -o jsonpath={..metadata.uid} -l release=app1-bf7534ac | tr " " "\n"
bf852ded-d7ea-11e8-ab5b-88c3b95d461f
bf85debb-d7ea-11e8-ab5b-88c3b95d461f
```
Only the first 8 chars differ, so my initial stab (actually using the first 8, not last 8) was complete luck. I'm thinking this is worth fixing before it's integrated into the SDK so we don't have to deal with users' release names changing when they upgrade their base helm operator image.

I like base64 idea, so I'll give that a go. That would give users 31 characters for the CR name, which seems like enough. I'll test what happens with a long CR name and add an additional check if necessary.

Reply:
Turns out k8s doesn't like uppercase letters in resource names, so the base64 approach didn't work:
```
ERRO[0003] failed to reconcile release for apiVersion=apache.org/v1alpha1 kind=Tomcat name=default/app1: install error: release app1-v3U0rNfqEeirW4jDuV1GHw failed: Service "app1-v3U0rNfqEeirW4jDuV1GHw-tomcat" is invalid: metadata.name: Invalid value: "app1-v3U0rNfqEeirW4jDuV1GHw-tomcat": a DNS-1035 label must consist of lower case alphanumeric characters or '-', start with an alphabetic character, and end with an alphanumeric character (e.g. 'my-name',  or 'abc-123', regex used for validation is '[a-z]([-a-z0-9]*[a-z0-9])?')
```

I switched to a base36 encoding (lowercase letters and digits 0-9), which works. It increases the length to 25 characters, so still not too bad.
```
INFO[0235] Reconciled release for apiVersion=apache.org/v1alpha1 kind=Tomcat name=default/app1 release=app1-41arrw2nar998e3vumtsue9lr
```

The resulting resource names are a bit long, but maybe still within the realm of reasonableness:
```
$ kubectl get all -l release=app1-41arrw2nar998e3vumtsue9lr
NAME                                                         READY   STATUS    RESTARTS   AGE
pod/app1-41arrw2nar998e3vumtsue9lr-tomcat-75b55c8855-9frcr   2/2     Running   0          5m
pod/app1-41arrw2nar998e3vumtsue9lr-tomcat-75b55c8855-qv966   0/2     Pending   0          5m

NAME                                            TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)   AGE
service/app1-41arrw2nar998e3vumtsue9lr-tomcat   ClusterIP   10.109.202.233   <none>        80/TCP    5m

NAME                                                    DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/app1-41arrw2nar998e3vumtsue9lr-tomcat   2         2         2            1           5m

NAME                                                               DESIRED   CURRENT   READY   AGE
replicaset.apps/app1-41arrw2nar998e3vumtsue9lr-tomcat-75b55c8855   2         2         1       5m
```

For release names that are too long, the operator prints a helpful error message:
```
ERRO[0000] failed to reconcile release for apiVersion=apache.org/v1alpha1 kind=Tomcat name=default/my-super-long-name-for-a-tomcat: install error: release name "my-super-long-name-for-a-tomcat-86jzwwlaodwvoiarjie4v87bz" exceeds max length of 53
```

As a separate bit of work, I don't think we're actually updating the CR status when we fail, so I'll add an issue for that.

WDYT?

Reply:
Yes let's follow up on the status update in a separate PR/issue.

And there might be a way to shorten the UUID further maybe by hashing it. But this is good enough for now.